File: hdar/archives/hdar_3/hdar/capsule/seal.py
"""Atomic capsule sealing.

Takes a workspace directory, ingests it into content-addressed storage,
creates a signed capsule manifest with lineage and receipt chain, and
writes the capsule as a single portable JSON file.
"""

from __future__ import annotations

import hashlib
import json
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Optional

from .store import ContentStore, WorkspaceManifest
from .identity import AgentIdentity, LineageEpoch, key_fingerprint, serialize_public_key
from .receipt import ReceiptChain, Receipt

from cryptography.hazmat.primitives.asymmetric.ed25519 import Ed25519PublicKey


@dataclass
class CapsuleManifest:
    """The sealed manifest inside a capsule.

    This is the agent's operational passport — not a complete copy
    of every machine it has visited.
    """
    # Identity
    agent_id: str
    agent_name: str
    epoch: Dict[str, Any]               # LineageEpoch.to_dict()
    parent_capsule_hash: Optional[str]   # None for genesis capsule

    # Model references (by digest, not embedded)
    model_digest: str = ""
    tokenizer_digest: str = ""
    inference_requirements: Dict[str, Any] = field(default_factory=dict)

    # Agent state
    objective: str = ""
    continuation_point: str = ""
    working_summary: str = ""

    # Workspace
    workspace_manifest: Optional[Dict[str, Any]] = None  # WorkspaceManifest.to_dict()

    # Capabilities
    capabilities: Dict[str, Any] = field(default_factory=dict)
    capability_note: str = ""

    # Secrets
    secret_references: List[Dict[str, str]] = field(default_factory=list)

    # Pending operations
    pending_operations: List[Dict[str, Any]] = field(default_factory=list)

    # Compatibility
    runtime_compatibility: Dict[str, Any] = field(default_factory=dict)

    # Restoration
    restoration_contract: str = "exact"  # exact | semantic | degraded

    # Receipts
    receipts: List[Dict[str, Any]] = field(default_factory=list)

    # Signature
    manifest_hash: str = ""
    signer_fingerprint: str = ""
    signature: str = ""
    sealed_at: float = 0.0

    def canonical_bytes(self) -> bytes:
        """Bytes that are signed (excludes signature, manifest_hash, sealed_at)."""
        d = self.to_dict()
        for k in ("signature", "manifest_hash", "sealed_at"):
            d.pop(k, None)
        return json.dumps(d, sort_keys=True, separators=(",", ":"), ensure_ascii=True).encode()

    def compute_hash(self) -> str:
        return hashlib.sha256(self.canonical_bytes()).hexdigest()

    def to_dict(self) -> dict:
        return {
            "agent_id": self.agent_id,
            "agent_name": self.agent_name,
            "epoch": self.epoch,
            "parent_capsule_hash": self.parent_capsule_hash,
            "model_digest": self.model_digest,
            "tokenizer_digest": self.tokenizer_digest,
            "inference_requirements": self.inference_requirements,
            "objective": self.objective,
            "continuation_point": self.continuation_point,
            "working_summary": self.working_summary,
            "workspace_manifest": self.workspace_manifest,
            "capabilities": self.capabilities,
            "capability_note": self.capability_note,
            "secret_references": self.secret_references,
            "pending_operations": self.pending_operations,
            "runtime_compatibility": self.runtime_compatibility,
            "restoration_contract": self.restoration_contract,
            "receipts": self.receipts,
            "manifest_hash": self.manifest_hash,
            "signer_fingerprint": self.signer_fingerprint,
            "signature": self.signature,
            "sealed_at": self.sealed_at,
        }

    @classmethod
    def from_dict(cls, d: dict) -> "CapsuleManifest":
        return cls(**{k: d.get(k) for k in [
            "agent_id", "agent_name", "epoch", "parent_capsule_hash",
            "model_digest", "tokenizer_digest", "inference_requirements",
            "objective", "continuation_point", "working_summary",
            "workspace_manifest", "capabilities", "capability_note",
            "secret_references", "pending_operations", "runtime_compatibility",
            "restoration_contract", "receipts", "manifest_hash",
            "signer_fingerprint", "signature", "sealed_at",
        ]})


class CapsuleSealer:
    """Seals a workspace into a portable, signed capsule.

    When a lease_manager is provided, sealing requires a valid fencing
    token. A stale runtime cannot publish a successor capsule.
    """

    def __init__(self, store: ContentStore, identity: AgentIdentity, lease_manager=None):
        self.store = store
        self.identity = identity
        self.lease_manager = lease_manager

    def seal(
        self,
        workspace_dir: Path,
        epoch: LineageEpoch,
        objective: str = "",
        continuation_point: str = "",
        working_summary: str = "",
        capabilities: Optional[dict] = None,
        capability_note: str = "",
        parent_capsule_hash: Optional[str] = None,
        model_digest: str = "",
        tokenizer_digest: str = "",
        inference_requirements: Optional[dict] = None,
        secret_references: Optional[list] = None,
        pending_operations: Optional[list] = None,
        runtime_compatibility: Optional[dict] = None,
        fencing_token: str = "",
    ) -> tuple[CapsuleManifest, ReceiptChain]:
        """Seal a workspace into a capsule.

        Returns (manifest, receipt_chain).

        When lease validation is enabled, a valid fencing token is required.
        A stale runtime cannot advance the authoritative lineage.
        """
        # 0. Validate fencing token if lease manager is present
        if self.lease_manager and fencing_token:
            if not self.lease_manager.validate_token(self.identity.agent_id, fencing_token):
                raise ValueError(
                    f"stale or invalid fencing token — "
                    f"this runtime's lease generation is no longer authoritative; "
                    f"cannot seal capsule"
                )
        # 1. Ingest workspace into content store
        ws_manifest = self.store.ingest_workspace(workspace_dir)

        # 2. Build receipt chain
        chain = ReceiptChain(
            agent_id=self.identity.agent_id,
            epoch_id=epoch.epoch_id,
            signing_key=self.identity.signing_key,
        )

        # 3. Append SEAL receipt
        seal_receipt = chain.append(
            receipt_type="SEAL",
            action="capsule_sealed",
            action_payload={
                "workspace_root_hash": ws_manifest.root_hash,
                "file_count": len(ws_manifest.files),
                "total_size": ws_manifest.total_size,
                "objective": objective,
            },
            state_root=ws_manifest.root_hash,
        )

        # 4. Build capsule manifest
        manifest = CapsuleManifest(
            agent_id=self.identity.agent_id,
            agent_name=self.identity.name,
            epoch=epoch.to_dict(),
            parent_capsule_hash=parent_capsule_hash,
            model_digest=model_digest,
            tokenizer_digest=tokenizer_digest,
            inference_requirements=inference_requirements or {},
            objective=objective,
            continuation_point=continuation_point,
            working_summary=working_summary,
            workspace_manifest=ws_manifest.to_dict(),
            capabilities=capabilities or {},
            capability_note=capability_note,
            secret_references=secret_references or [],
            pending_operations=pending_operations or [],
            runtime_compatibility=runtime_compatibility or {},
            restoration_contract="exact",
            receipts=chain.to_list(),
            signer_fingerprint=self.identity.fingerprint,
            sealed_at=time.time(),
        )

        # 5. Sign manifest with Ed25519
        manifest.manifest_hash = manifest.compute_hash()
        manifest.signature = self.identity.sign(manifest.canonical_bytes()).hex()

        return manifest, chain

    def write_capsule(self, manifest: CapsuleManifest, output_path: Path):
        """Write a capsule manifest as a JSON file."""
        output_path.parent.mkdir(parents=True, exist_ok=True)
        output_path.write_text(
            json.dumps(manifest.to_dict(), indent=2, ensure_ascii=True)
        )

    def verify_manifest(self, manifest: CapsuleManifest, public_key: Ed25519PublicKey) -> bool:
        """Verify the manifest Ed25519 signature."""
        try:
            public_key.verify(
                bytes.fromhex(manifest.signature),
                manifest.canonical_bytes(),
            )
            return True
        except Exception:
            return False

    def verify_capsule_file(self, capsule_path: Path, public_key: Ed25519PublicKey) -> bool:
        """Load and verify a capsule file's manifest hash and signature."""
        data = json.loads(capsule_path.read_text())
        manifest = CapsuleManifest.from_dict(data)
        expected_hash = manifest.compute_hash()
        if manifest.manifest_hash != expected_hash:
            return False
        return self.verify_manifest(manifest, public_key)

File: hdar/hdar_core/cpp/src/seal.cpp
#include "hdar/seal.hpp"
#include "hdar/lease.hpp"

namespace hdar {

// ── CapsuleManifest ───────────────────────────────────────────

std::vector<uint8_t> CapsuleManifest::canonical_bytes() const {
    JsonValue d = to_json();
    // Remove signature, manifest_hash, sealed_at
    d.object_val.erase("signature");
    d.object_val.erase("manifest_hash");
    d.object_val.erase("sealed_at");
    return canonical_json_bytes(d);
}

std::string CapsuleManifest::compute_hash() const {
    return sha256_hex(canonical_bytes());
}

JsonValue CapsuleManifest::to_json() const {
    JsonValue v = JsonValue::object();
    v["agent_id"] = JsonValue::string(agent_id);
    v["agent_name"] = JsonValue::string(agent_name);
    v["epoch"] = epoch;
    if (parent_capsule_hash)
        v["parent_capsule_hash"] = JsonValue::string(*parent_capsule_hash);
    else
        v["parent_capsule_hash"] = JsonValue::null();
    v["model_digest"] = JsonValue::string(model_digest);
    v["tokenizer_digest"] = JsonValue::string(tokenizer_digest);
    v["inference_requirements"] = inference_requirements;
    v["objective"] = JsonValue::string(objective);
    v["continuation_point"] = JsonValue::string(continuation_point);
    v["working_summary"] = JsonValue::string(working_summary);
    if (workspace_manifest)
        v["workspace_manifest"] = *workspace_manifest;
    else
        v["workspace_manifest"] = JsonValue::null();
    v["capabilities"] = capabilities;
    v["capability_note"] = JsonValue::string(capability_note);
    v["secret_references"] = secret_references;
    v["pending_operations"] = pending_operations;
    v["runtime_compatibility"] = runtime_compatibility;
    v["restoration_contract"] = JsonValue::string(restoration_contract);
    v["receipts"] = receipts;
    v["manifest_hash"] = JsonValue::string(manifest_hash);
    v["signer_fingerprint"] = JsonValue::string(signer_fingerprint);
    v["signature"] = JsonValue::string(signature);
    v["sealed_at"] = JsonValue::number(sealed_at);
    return v;
}

CapsuleManifest CapsuleManifest::from_json(const JsonValue& d) {
    CapsuleManifest m;
    m.agent_id = d.get("agent_id").string_val;
    m.agent_name = d.get("agent_name").string_val;
    m.epoch = d.get("epoch");
    const auto& pch = d.get("parent_capsule_hash");
    if (pch.type == JsonValue::Type::String)
        m.parent_capsule_hash = pch.string_val;
    m.model_digest = d.get("model_digest").string_val;
    m.tokenizer_digest = d.get("tokenizer_digest").string_val;
    m.inference_requirements = d.get("inference_requirements");
    m.objective = d.get("objective").string_val;
    m.continuation_point = d.get("continuation_point").string_val;
    m.working_summary = d.get("working_summary").string_val;
    const auto& wm = d.get("workspace_manifest");
    if (wm.type == JsonValue::Type::Object)
        m.workspace_manifest = wm;
    m.capabilities = d.get("capabilities");
    m.capability_note = d.get("capability_note").string_val;
    m.secret_references = d.get("secret_references");
    m.pending_operations = d.get("pending_operations");
    m.runtime_compatibility = d.get("runtime_compatibility");
    m.restoration_contract = d.get("restoration_contract").string_val;
    if (m.restoration_contract.empty()) m.restoration_contract = "exact";
    m.receipts = d.get("receipts");
    m.manifest_hash = d.get("manifest_hash").string_val;
    m.signer_fingerprint = d.get("signer_fingerprint").string_val;
    m.signature = d.get("signature").string_val;
    m.sealed_at = d.get("sealed_at").double_val;
    return m;
}

// ── CapsuleSealer ─────────────────────────────────────────────

CapsuleSealer::CapsuleSealer(ContentStore& store, AgentIdentity& identity,
                             LeaseManager* lm)
    : store_(store), identity_(identity), lease_manager_(lm) {}

std::pair<CapsuleManifest, ReceiptChain> CapsuleSealer::seal(
    const std::string& workspace_dir,
    const LineageEpoch& epoch,
    const std::string& objective,
    const std::string& continuation_point,
    const std::string& working_summary,
    const JsonValue& capabilities,
    const std::string& capability_note,
    const std::optional<std::string>& parent_capsule_hash,
    const std::string& model_digest,
    const std::string& tokenizer_digest,
    const JsonValue& inference_requirements,
    const JsonValue& secret_references,
    const JsonValue& pending_operations,
    const JsonValue& runtime_compatibility,
    const std::string& fencing_token) {

    // 0. Validate fencing token if lease manager is present
    if (lease_manager_ && !fencing_token.empty()) {
        if (!lease_manager_->validate_token(identity_.agent_id, fencing_token)) {
            throw std::runtime_error(
                "stale or invalid fencing token — this runtime's lease generation "
                "is no longer authoritative; cannot seal capsule");
        }
    }

    // 1. Ingest workspace
    auto ws_manifest = store_.ingest_workspace(workspace_dir);

    // 2. Build receipt chain
    ReceiptChain chain(identity_.agent_id, epoch.epoch_id, identity_.signing_key);

    // 3. Append SEAL receipt
    JsonValue seal_payload = JsonValue::object();
    seal_payload["workspace_root_hash"] = JsonValue::string(ws_manifest.root_hash);
    seal_payload["file_count"] = JsonValue::integer(static_cast<int64_t>(ws_manifest.files.size()));
    seal_payload["total_size"] = JsonValue::integer(static_cast<int64_t>(ws_manifest.total_size));
    seal_payload["objective"] = JsonValue::string(objective);

    chain.append("SEAL", "capsule_sealed", seal_payload, ws_manifest.root_hash);

    // 4. Build manifest
    CapsuleManifest manifest;
    manifest.agent_id = identity_.agent_id;
    manifest.agent_name = identity_.name;
    manifest.epoch = epoch.to_json();
    manifest.parent_capsule_hash = parent_capsule_hash;
    manifest.model_digest = model_digest;
    manifest.tokenizer_digest = tokenizer_digest;
    manifest.inference_requirements = inference_requirements;
    manifest.objective = objective;
    manifest.continuation_point = continuation_point;
    manifest.working_summary = working_summary;
    manifest.workspace_manifest = ws_manifest.to_json();
    manifest.capabilities = capabilities;
    manifest.capability_note = capability_note;
    manifest.secret_references = secret_references;
    manifest.pending_operations = pending_operations;
    manifest.runtime_compatibility = runtime_compatibility;
    manifest.restoration_contract = "exact";
    manifest.receipts = chain.to_json_array();
    manifest.signer_fingerprint = identity_.fingerprint();
    manifest.sealed_at = epoch_seconds();

    // 5. Sign manifest
    manifest.manifest_hash = manifest.compute_hash();
    manifest.signature = identity_.sign_hex(manifest.canonical_bytes());

    return {manifest, chain};
}

void CapsuleSealer::write_capsule(const CapsuleManifest& manifest,
                                   const std::string& output_path) const {
    auto parent = std::filesystem::path(output_path).parent_path();
    if (!parent.empty())
        std::filesystem::create_directories(parent);
    write_string_to_file(output_path, canonical_json(manifest.to_json()));
}

bool CapsuleSealer::verify_manifest(const CapsuleManifest& manifest,
                                     const PublicKey& pk) const {
    return pk.verify_hex(manifest.canonical_bytes(), manifest.signature);
}

bool CapsuleSealer::verify_capsule_file(const std::string& capsule_path,
                                         const PublicKey& pk) const {
    std::string content = read_file_to_string(capsule_path);
    if (content.empty())
        return false;
    try {
        JsonValue json = parse_json(content);
        auto manifest = CapsuleManifest::from_json(json);
        return verify_manifest(manifest, pk);
    } catch (const std::exception&) {
        return false;
    }
}

} // namespace hdar
